1.0.1 / 2023-09-18

yermulnik opened the pull request:
- Add installation instruction to [`README`](./README)
- `usage()`: use variable instead of hardcoded tool name
- `halt-error()`: print diagnostic only if there are still items in `$@`
- Add `-h | --help | --usage` into cmdline argument parser loop
- Workaround use of `insteadOf` GIT CLI config option by replacing `git remote get-url "${remote}"`
  with `git config --get remote."${remote}".url`

yermulnik committed:
1.0.1 / 2023-09-18

- Add installation instruction to [`README`](./README)
- `usage()`: use variable instead of hardcoded tool name
- `halt-error()`: print diagnostic only if there are still items in `$@`
- Add `-h | --help | --usage` into cmdline argument parser loop
- Workaround use of `insteadOf` GIT CLI config option by replacing `git remote get-url "${remote}"`
  with `git config --get remote."${remote}".url`

yermulnik committed:
Fix typo in CHANGELOG

halostatue commented on gh-merge-upstream:
```diff
@@ -64,7 +64,7 @@ halt-error() {
   fi
 
-  echo >&2 "${bn0}: $*"
+  [[ $@ ]] && echo >&2 "${bn0}: $*"
```
I prefer `(($#)) &&` over `[[ $@ ]]`, but I also think we should print *something* when there are unknown errors.

```bash
(($#) || set -- error
echo >&2 "${bn0}: $*"
```

That will fill `$@` with `error` so we get `gh-merge-upstream: error` at least. Combined with the change of just using `usage; exit 0` for `-h`, we have a case where simply calling `halt-error` is sufficient.

yermulnik replied:
I think it's sort of anti-pattern to check whether var is empty or not using arithmetic evaluation of its length. Use of conditional expression to test whether string is not zero is what `man 1 bash` describes under `CONDITIONAL EXPRESSIONS` section:
![image](https://github.com/halostatue/gh-merge-upstream/assets/1274789/5e2fccb1-5e9b-43e5-8c39-9f63d5abc3a5)

halostatue replied:
We’re not using the length of a string, but the length of an array. `$#` is the number of elements in the `$@` array. The syntax is easier to see with normal array variables, but `$#` is `${#@}`, which is not `${#1}`.

The difference is much more visible with normal array variables:

```console
f() {
  x=("$@")
  echo param count: $# == ${#@} == ${#x[@]}
  echo first arg length: ${#1} == ${#x} == ${#x[0]}
}

f aa b ccc

param count: 3 == 3 == 3
first arg length: 2 == 2 == 2
```

Since we're checking array length, I prefer `(($#))` over `[[ $@ ]]`.

yermulnik replied:
In this context what you're willing to do in the code is to check the emptiness of the value (no matter whether this is a string or an array, which in Bash is not that kind of true `array` that you might be thinking it is) to do the only action: to decide whether to fallback to a default value by overriding positional parameters array OR to print what actually has left in original array of positional args. Consider the difference:
```shell
> EMPTY_VALS=("")
> echo ${#EMPTY_VALS[@]}
1
> ((${#EMPTY_VALS[@]})) && echo "<The error is: ${EMPTY_VALS[*]}>"
<The error is: >
> [[ ${EMPTY_VALS[@]} ]] && echo "<The error is: ${EMPTY_VALS[*]}>"
>
```
Note that the very last command checks the emptiness of the value and does not provide output on an **empty** value, whilst arithmetic evaluation solution designates the **difference** between these two by outputting visually empty value, and also this approach inadvertently (and to the matter of fun in the context of this PR) allows arrays with negative number of elements (obviously absolutely not sure whether this is possible in Bash) 😺 

PS: I'm starting to like how this discussion became not just trivial 🤝

halostatue replied:
This is one of those corner cases where I remember how much I hate Bash’s inconsistencies.

```console
$ x="a b c"
$ [[ $x ]] && echo true
true
$ f() { echo $#; }
$ f $x
3
$ [[ a b c ]] && echo true
bash: syntax error near unexpected token `a'
$ [[ -n $x ]] && echo true
true
$ [[ -n a b c ]] && echo true
bash: syntax error in conditional expression
```

Inasmuch as this is a script-internal function which would not be passed an empty string, I would *prefer* `(($#))`, but your assessment is correct on `(($#))` detecting the difference between `halt-error` and `halt-error ""`. I will accept this as written, but understand that I’m looking at it with much side-eye because it's just bad behaviour. I would almost prefer `[[ -z "$*" ]] && set -- error` over `[[ $@ ]] || set -- error`.

If it annoys me too much should there be much more development with this, I might even rewrite that part to do something like:

```bash
local message
message="$*"

[[ -z "${message}" ]] && message=error

echo >&2 "${bn0}: ${message}"
```

Then again, if there is more development to happen with this, I’ll need to set up GitHub actions with `shfmt` and `shellcheck` and maybe even `prettier` for the Markdown.

yermulnik replied:
Those are not inconsistencies from my point of view. They are "features" of the shell interpreter to support common set of operators and functions natively to allow "conveniency" developing scripts (as in batch-files) rather than programming-language based applications.
Also I don't think giving up using modern Bash features is nice idea (as in `[[ -n $var ]]` or even `[[ -n "$var" ]]` instead of `[[ ${var} ]]`. If there's no decent reason to support ancient versions of Bash and/or to be strictly POSIX-compliant, then why would you want to decline using modern Bash =)

> if there is more development to happen with this

Despite I'm just a bypasser who noticed new GH extension with potentially useful use case and shared backported codebits that resolved my local needs, I'd be happy to participate in the development by sharing my expertise via PR reviews or alike.

I appreciate you, accepting my contribution, and pretty interesting discussion 🤝

halostatue commented on gh-merge-upstream:
```diff
@@ -170,6 +170,9 @@ while (($#)); do
     shift
     ;;
+  -h | --help | --usage)
+    halt-error --usage
```
I had meant to add this, but it should be

```bash
-h | --help | --usage)
  usage
  exit 0
  ;;
```

When deliberately asking for help, the exit code should be 0.

yermulnik replied:
Makes sense. Updating.

halostatue commented on gh-merge-upstream:
```diff
@@ -201,7 +204,9 @@ if [[ "${in_git}" ]]; then
 
   for remote in "${remotes[@]}"; do
-    remote_urls["${remote}"]="$(git remote get-url "${remote}")"
+    #remote_urls["${remote}"]="$(git remote get-url "${remote}")"
```
This line can be removed.

yermulnik replied:
✔️

halostatue commented on gh-merge-upstream:
```diff
@@ -201,7 +204,9 @@ if [[ "${in_git}" ]]; then
   for remote in "${remotes[@]}"; do
-    remote_urls["${remote}"]="$(git remote get-url "${remote}")"
+    #remote_urls["${remote}"]="$(git remote get-url "${remote}")"
+    # workaround "insteadOf" by querying local config
```
I would prefer a slightly different comment:

```
# workaround "url.<base>.insteadOf" mapping by querying local config
```

I had to look up `insteadOf` because I haven’t used it in…forever.

This would imply updating the CHANGELOG record, too.

yermulnik replied:
> I had to look up `insteadOf` because I haven’t used it in…forever.

Yep, I hadn't too, until my Company signed for Github EMU 🤷🏻 
I'm addressing this suggestion too. Thanks.

halostatue commented on CHANGELOG.md:
```diff
@@ -1,5 +1,13 @@
 # gh-merge-upstream
 
+## 1.0.1 / 2023-09-18
+- Add installation instruction to [`README`](./README.md)
```
This is a nit, but please add a blank line after headings. I generally run `prettier` for more-or-less default Markdown formatting.

yermulnik replied:
✔️

yermulnik replied:
Btw you may also want to make use of Git tags to mark new versions of this extension.

halostatue reviewed (changes requested):
Thank you very much for your contribution! It was very unexpected for a tool that is only a couple of days old.

I have a couple of changes I would like made, but we can get this merged very quickly with those.

I will ask you to make one other change to the README. Please add the following block at the end:

```markdown
## Contributors

- Austin Ziegler (@halostatue) created gh-merge-upstream.
- George L. Yermulnik (@yermulnik)
```

If you prefer to just use your GitHub handle, that is fine, too, but I do like recognizing contributors to projects I create or manage.

yermulnik committed:
Addtress @halostatue (Austin Ziegler) comments

yermulnik commented:
@halostatue I addresses most of your comments. Thank you for the thorough review.
Let me know what do you think on `(($#))` vs `[[ @ ]]` and overall PR after last commit. Thanks.

yermulnik commented:
By the way... it looks like I'm going to utilize this GH extension to try and sync my fork with upstream once this PR is merged 😺

yermulnik committed:
Linkify user mentions in README

halostatue commented:
I’ve replied to the one item listed and I would still like to see `(($#))` over `[[ $@ ]]`.
